File: backend/app/models/traffic.py
from datetime import datetime
from pydantic import BaseModel, Field


class TrafficPoint(BaseModel):
    time: datetime
    bytes_sum: float = 0.0
    packets_sum: float = 0.0
    flows_count: float = 0.0
    proto: str | None = None


class TrafficSummaryResponse(BaseModel):
    start: datetime
    stop: datetime
    total_bytes: float = 0.0
    total_packets: float = 0.0
    total_flows: float = 0.0
    points: list[TrafficPoint] = Field(default_factory=list)


class TopEndpoint(BaseModel):
    ip: str
    bytes_sum: float = 0.0
    packets_sum: float = 0.0
    flows_count: float = 0.0


class TopEndpointsResponse(BaseModel):
    start: datetime
    stop: datetime
    endpoints: list[TopEndpoint] = Field(default_factory=list)


class ProtocolOption(BaseModel):
    proto: str
    label: str


class ProtocolBucket(BaseModel):
    proto: str
    label: str
    bytes_sum: float = 0.0
    packets_sum: float = 0.0
    flows_count: float = 0.0


class ProtocolDistributionResponse(BaseModel):
    start: datetime
    stop: datetime
    protocols: list[ProtocolBucket] = Field(default_factory=list)


class FlowRecord(BaseModel):
    time: datetime
    src_ip: str
    src_port: int = 0
    dst_ip: str
    dst_port: int = 0
    proto: str
    proto_label: str
    bytes: float = 0.0
    packets: float = 0.0
    direction: str


class FlowsResponse(BaseModel):
    start: datetime
    stop: datetime
    offset: int
    limit: int
    flows: list[FlowRecord] = Field(default_factory=list)

File: frontend/src/types/traffic.ts
export interface TrafficPoint {
  time: string;
  bytes_sum: number;
  packets_sum: number;
  flows_count: number;
  proto: string | null;
}

export interface TrafficSummaryResponse {
  start: string;
  stop: string;
  total_bytes: number;
  total_packets: number;
  total_flows: number;
  points: TrafficPoint[];
}

export interface TopEndpoint {
  ip: string;
  bytes_sum: number;
  packets_sum: number;
  flows_count: number;
}

export interface TopEndpointsResponse {
  start: string;
  stop: string;
  endpoints: TopEndpoint[];
}

export interface ProtocolBucket {
  proto: string;
  label: string;
  bytes_sum: number;
  packets_sum: number;
  flows_count: number;
}

export interface ProtocolDistributionResponse {
  start: string;
  stop: string;
  protocols: ProtocolBucket[];
}

export interface ProtocolOption {
  proto: string;
  label: string;
}

export interface TrafficFilters {
  start: string;
  stop: string;
  proto?: string;
  src_ip?: string;
  dst_ip?: string;
  src_port?: number;
  dst_port?: number;
}

export interface FlowRecord {
  time: string;
  src_ip: string;
  src_port: number;
  dst_ip: string;
  dst_port: number;
  proto: string;
  proto_label: string;
  bytes: number;
  packets: number;
  direction: 'inbound' | 'outbound' | 'internal' | 'external';
}

export interface FlowsResponse {
  start: string;
  stop: string;
  offset: number;
  limit: number;
  flows: FlowRecord[];
}
